lib/Dialect/MSFT/MSFTPasses.cpp

```diff
@@ -502,6 +502,122 @@ static StringRef getOperandName(OpOperand &oper, const hw::SymbolCache &syms,
+
+      auto newLeaf = hw::InnerRefAttr::get(newModule, newNameAttr);
+      newPath.push_back(newLeaf);
+
+      // Update the path on the GlobalRefOp.
+      auto newPathAttr = ArrayAttr::get(op->getContext(), newPath);
+      globalRefOp.namepathAttr(newPathAttr);
+    }
+  }
+}
+
+/// Helper to update GlobalRefops after referenced ops are pushed down.
+static void pushDownGlobalRefs(Operation *op) {
+  if (auto globalRefs = op->getAttrOfType<ArrayAttr>("circt.globalRef")) {
+    for (auto globalRef : globalRefs.getAsRange<hw::GlobalRefAttr>()) {
+      // Resolve the GlobalRefOp and get its path.
+      auto rootModule = op->getParentOfType<ModuleOp>();
+      auto refSymbol = globalRef.getGlblSym();
+      auto globalRefOp = rootModule.lookupSymbol<hw::GlobalRefOp>(refSymbol);
+      auto oldPath = globalRefOp.namepath().getAsRange<hw::InnerRefAttr>();
+
+      // Get the module containing the partition and the partition's name.
+      auto partAttr = op->getAttrOfType<SymbolRefAttr>("targetDesignPartition");
+      auto partMod = partAttr.getRootReference();
+      auto partName = partAttr.getLeafReference();
+
+      // Resolve the DesignPartitionOp and get its verilog name.
+      auto mod = op->getParentOfType<MSFTModuleOp>();
+      StringAttr partModName;
+      for (auto part : mod.getOps<DesignPartitionOp>()) {
+        if (part.sym_nameAttr() == partName) {
+          partModName = part.verilogNameAttr();
+          break;
+        }
+      }
+      assert(partModName);
+
+      // Construct a new path starting from the old path.
+      SmallVector<Attribute> newPath(oldPath);
+      auto lastElement = newPath.pop_back_val().cast<hw::InnerRefAttr>();
+
+      // If the last element in the path is not pointing at the module
+      // containing the partition, this GlobalRefOp has already been updated.
+      if (lastElement.getModule() != partMod)
+        continue;
+
+      // Split the last element in the path to add the partition.
+      auto partRef = hw::InnerRefAttr::get(partMod, partName);
+      auto leafRef = hw::InnerRefAttr::get(partModName, lastElement.getName());
+      newPath.push_back(partRef);
+      newPath.push_back(leafRef);
+
+      // Update the path on the GlobalRefOp.
+      auto newPathAttr = ArrayAttr::get(op->getContext(), newPath);
+      globalRefOp.namepathAttr(newPathAttr);
+
+      // Find the instance of the design partition.
+      InstanceOp partInstance;
+      for (auto instOp : mod.getOps<InstanceOp>()) {
+        if (instOp.sym_nameAttr() == partName) {
+          partInstance = instOp;
+          break;
+        }
+      }
+      assert(partInstance);
+
+      // Ensure the instance has this GlobalRefAttr, or collect any existing
+      // global refs if not.
+      SmallVector<Attribute> newGlobalRefs;
+      auto partRefs = partInstance->getAttrOfType<ArrayAttr>("circt.globalRef");
+      if (partRefs) {
+        for (auto ref : partRefs.getAsRange<hw::GlobalRefAttr>()) {
```

Comment:
Is there any way to avoid this loop? Could you accept a `DenseSet&` global refs to set on each instance, add the global ref then set them on the instance in the caller?

Reply:
This would also avoid creating a new `ArrayAttr` (for `newRefsAttr`) per-op which would be overwritten pretty much immediately (if I'm reading the code correctly).